### DISCUSSION TITLE
关于 ConnectButton 的 buttonProps 属性

### DISCUSSION THREAD

**Discussion by kiner-tang** [ant-design/ant-design-web3#60] [Ideas]
目前，我们的 ConnectButton 是基于 antd 的 Button 进行的上层封装，在使用过程中，可能需要使用到 antd 的 Button 的各种原生属性，当前我们的设计是这样的：
<img width="837" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/10286961/18a7b40a-3779-4edc-b166-501e2aaf1937">
这样设计有个好处是可以直接像使用 antd 的 Button 一样使用 ConnectButton，但也可能带来一些副作用，例如 antd 中 Button 属性增加了一个新的属性，刚好跟我们 ConnectButton 自定义的属性命名重名的话就会出现了冲突。以及如果使用这个设计的话，我们在文档中不方便描述那些原始属性。
因此，这边想参考 pro-components 中的设计，如：[modalProps](https://procomponents.ant.design/components/modal-form#api) 属性，这样，我们在文档中只需要简单的导引过去即可：
<img width="1050" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/10286961/d2a24ddc-1b2e-410c-a8b9-d061a347ffe7">
<img width="731" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/10286961/98f530d2-a8dd-42f5-b42e-1b1295d17fa9">
并且可以避免上述的命名冲突的问题。

相关改动已经有 PR，可以到：https://github.com/ant-design/ant-design-web3/pull/59 看看

**Comment by RedJue** [comment]
赞成第二种，相对解耦，因为我们不是基础组件，一些原生属性理论上应该只提供接口暴露。

**Comment by yutingzhao1991** [comment]
我觉得不需要， 这个组件本来就叫 ConnectButton，上面再有一个 buttonProps 太奇怪了。倒是可以有一个 `modalProps`。

冲突的问题感觉概率很小，即便是出现也可以换一个别名。

就好像 ProCard 继承了 Card 的属性一样，如果参考 antd pro 的话就更不应该修改了。

**Reply by kiner-tang** [reply]
我感觉后面我们的 ConnectButton 增加一个 type 的属性还是有可能的，但这个 type 不一定跟原 antd Button 的 type 是同样的含义，如果 ConnectButton 换一个别名的话，感觉多少有点累赘。此外，如果我们原始的属性跟 ConnectButton 的属性混在一起，感觉不太好维护，内部使用时如果嵌套层级比较深了，容易搞混。再者，如果不单独拆开，文档也不好描述。🤔

**Reply by yutingzhao1991** [reply]
ConnectButton 的 type 就是 Button 的 type 看上去也没什么毛病，如果是其他类型，比如连接类型那应该叫做 `ConnectType`。

不过我倒是在想，如果 `ConnectButton` 换一个名字，比如叫 `Connector` 那就没问题了，就可以用 `buttonProps`，不过要不要这么做还是得看这个组件设计上的主体是什么，如果整个设计都是围绕 `Button` 的话我觉得还是叫 `ConnectButton` 好，但是如果不一定是 Button，比如某种情况可能是 Dropdown 或者别的那可能就要换 `Connector` 了。

所以就看看设计稿以及我们对这个组件的定义上看是不是一直都会有 Button 来决定。

**Reply by kiner-tang** [reply]
感觉是不是可以参考 antd 中的变种组件的设计形式：
<img width="768" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/10286961/fca13066-dc33-43d1-8c27-2bd6e9fabd4b">
<img width="947" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/10286961/710d5ba2-86eb-4726-b194-229d562eb8ef">

感觉以后还是有可能有各种连接创建器的变种的，刚开始就自我设限成 Button 好像不太好，我们是不不是整体的链接创建器叫：Connector,然后目前只有 Button 的 Connector 的话，可以用 Connector.Button 的方式。

**Reply by yutingzhao1991** [reply]
嗯 我们再具体看看吧，可以按照现在设计稿的样式设计个完整的 API 出来看看。

**Reply by RedJue** [reply]
如果有变体的情况还是需要考虑多一些，毕竟开出去了就不好收了🤔

**Reply by kiner-tang** [reply]
目前设计稿，至少纯 Button 和 Dropdown 感觉就可以看成两种不同的变体：
<img width="1399" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/10286961/0491828c-7bdc-42f1-825d-8475a012dacb">

**Reply by RedJue** [reply]
那名字可以换一下，用连接器可能更适合，API 参考 antd 的变体形式

**Reply by RedJue** [reply]
Button 形式可以作为默认展现，可以不用xx.Button，其他变体可以跟上后缀

**Reply by kiner-tang** [reply]
@yutingzhao1991 觉得如何，如果这个方案没问题，我再出一个详细的 RFC

**Reply by yutingzhao1991** [reply]
嗯看上去OK

**Comment by yutingzhao1991** [comment]
这个组件很重要，得 work 起来了，可以按照 https://web3.ant.design/zh-CN/guide/contributing 拆分下任务，这样可以更多人一起进来协作。

@kiner-tang @jeasonstudio 看看你们谁先做下一个模块拆分的设计，先按照新的设计思路把架子搭起来，尽快可以推动可以开发的任务拆分出来。这样我这边也可以一起来看看。

**Reply by kiner-tang** [reply]
好的，我这两天捋一捋，再跟 @jeasonstudio 一起出一个 RFC 出来大家讨论一下

**Reply by yutingzhao1991** [reply]
嗯，可以先把大概的组件架子和内部模块的接口定下来，这样后面可以大家一起来搞。

**Comment by jeasonstudio** [comment]
`connector` 组件本质是 `button` 的组合，但在当前的垂直场景下部分 props 可以考虑做裁剪，一些基本的设计如下：
@kiner-tang 

```ts

export interface Chain {} // MOCK
export interface Account {} // MOCK

// Universal Props
// danger 看起来不需要
// href/htmlType/target
export interface UniversalConnectorProps
  extends Exclude<ButtonProps, 'danger' | 'href' | 'htmlType' | 'target'> {
  // 这里应该继承 Web3ConfigProvider 上定义的部分通用 props
  // 比如 onRequestAccounts/onRequestTokenMetadata 等
  // 这些与 UI 无关，暂时省略
}

// 链切换的 button/dropdown props, 无论展示形式如何都应该用 button 做 trigger
export interface ChainSwitcherProps extends UniversalConnectorProps {
  // 允许的链列表（这个只能受控）
  // 但如果不传我们应该要默认只有一个 eth mainnet
  chains?: Chain[];
  onChainChange?: (chain: Chain) => void;

  // 由于要兼容不同的弹出形式（dropdown 和 modal），这里不继承某一个组件的 props
  open?: boolean;
  onOpenChange?: (open: boolean) => void;
  destroyOnHide?: boolean;
}

// 未连接状态下的按钮
export interface UnconnectedConnectorProps extends UniversalConnectorProps {
  // 未连接状态按钮应该包含 `ConnectModal` 组件的 props
  // 这里预留出来
}

// 已连接状态下的按钮
export interface ConnectedConnectorProps extends UniversalConnectorProps {
  // 当前账户，如果不传则为非受控，传了按受控处理
  account?: Account;
  onAccountChange?: (account: Account) => void;

  // 是否展示余额
  showBalance?: boolean;

  // 是否展示地址信息(包括地址和 ens)
  showAddress?: 'full' | 'ellipsis' | 'ens' | false;

  // 是否展示头像
  showAvatar?: boolean;

  // 同样要兼容不同的弹出形式（dropdown 和 modal）
  // 本期暂时不支持扩展 dropdown 的 menus
  open?: boolean;
  onOpenChange?: (open: boolean) => void;
  destroyOnHide?: boolean;
}

// 对多种状态的组合组件，需要包含上述用到的几个组件的 props
export interface ConnectorProps
  extends UniversalConnectorProps,
    Pick<ChainSwitcherProps, 'chains' | 'onChainChange'>,
    Pick<
      ConnectedConnectorProps,
      'account' | 'onAccountChange' | 'showBalance' | 'showAddress' | 'showAvatar'
    > {
  // 参考 antd 命名规范选择了 showChainSwitcher
  showChainSwitcher?: boolean;
  // ChainSwitcherProps rename
  chainSwitcherOpen?: boolean;
  onchainSwitcherOpenChange?: (open: boolean) => void;
  destroyOnchainSwitcherHide?: boolean;

  // ConnectedConnectorProps rename
  connectorOpen?: boolean;
  onConnectorOpenChange?: (open: boolean) => void;
  destroyOnConnectorHide?: boolean;
}

```

**Reply by yutingzhao1991** [reply]
UnconnectedConnectorProps 这个是在哪里用的？看上面好像只做了定义，没有地方透出。

**Reply by kiner-tang** [reply]
> // 链切换的 button/dropdown props, 无论展示形式如何都应该用 button 做 trigger

感觉这个是不是不必要限制得那么死，其实后期或许会有自定义 trigger 的方式满足更加灵活的场景，不一定要限制在 button 里面

**Reply by jeasonstudio** [reply]
`UnconnectedConnector` 的功能被 `Connector` 包含掉，这里是因为没有特殊的 props 所以没体现在 `ConnectorProps ` 中

**Reply by jeasonstudio** [reply]
目前的大部分 UI props 都依赖 Button 的实现，如果不用 Button 做 trigger 会很难做适配，而且短期来看应该也不会有太灵活的扩展需求

**Reply by yutingzhao1991** [reply]
这个方案对应的用法是怎么样的？是下面这样直接吧 props 通过 Connector 暴露吗：

```tsx
<Connector {...props} />
```

**Comment by yutingzhao1991** [comment]
我有一个想法大概是：

```tsx
import { Connector } from '@ant-design/web3';

export default () => {
  return (
    <Connector
     /* 链相关的一些 props */
     accounts={[]}
     showBalance
     onRequestAccount={() => {}}
     chains={[]}
     onChainChanged={}

     /* 里面用到的各种组件暴露出来用户自定义 UI 的 props */
     connectButtonProps={/* ... */}
     connectModalProps={/* ... */}
     switchChainDropdownProps={/* ... */}
    />
  );
}

```

**Reply by kiner-tang** [reply]
感觉这样 Connector 就太大了，个人还是比较倾向于变种组件的形式：
```tsx
export default () => {
  return (
     <>
          <Connector>text</Connector>
          <Connector.Button>Button</Connector.Button>
          <Connector.Dropdown>Dropdown</Connector.Dropdown>
     </>
  )
}
```

**Reply by yutingzhao1991** [reply]
但是 Dropdown 并不是一种 Connector 的形式，只是说在出现多链需要选择的时候需要用到 Dropdown

**Reply by kiner-tang** [reply]
主要是如果把所有的属性都混在一起的话，以后 Connector 会变得特别大，逻辑糅杂，不太好处理

**Reply by yutingzhao1991** [reply]
Connector 直接暴露的 props 只是必须的，其它的 props 只是暴露出来给用户自定义的，其实如果没有需求也可以不用暴露出来。

**Comment by yutingzhao1991** [comment]
其实 UI 上感觉有点像 `Dropdown.Button` 这个 https://ant.design/components/dropdown-cn#dropdownbutton。

<img width="228" alt="image" src="https://github.com/ant-design/ant-design-web3/assets/1061968/f7ddd868-340c-46cc-bc5c-f22f026d0658">

**Reply by yutingzhao1991** [reply]
如果从这个角度看的话其实也不需要太纠结，就先搞一个 `Connector.Button`，以 Button 为主体，暴露需要的 props 即可：

```tsx
export default () => {
  return (
    <Connector.Button chain={[]} onClick={() => {}} />
  );
}
```

也可以这样：

```tsx
export default () => {
  return (
    <Connector chain={[]}>
       <Button />
    </Connector>
  );
}
```

**Reply by kiner-tang** [reply]
恩恩，我刚开始想的也是 Connector 作为一个包装容器，而子组件作为触发器可以让用户自定义，只是我们可以提供一些常见的变种，如：Connector.Button 或 Connector.Dropdown 之类的

**Reply by yutingzhao1991** [reply]
Button 应该还是最基础的，其次就是有多链切换的类似 `Dropdown.Button` 的形式。如果用户不想用 Button，比如想用 icon 那就这样：

```
export default () => {
  return (
    <Connector chain={[]}>
       <XxxIcon />
    </Connector>
  );
}
```

**Reply by yutingzhao1991** [reply]
` Connector.Dropdown` 应该不太需要，Dropdown 只是用来下来选择链的。不过我这样的思路可以， @jeasonstudio  觉得如何？

**Comment by yutingzhao1991** [comment]
除了组件，还要支持 hooks 的方式：

```tsx
import { useConnector } from '@ant-design/web3';

export default () => {
  const { connect } = useConnector({
    chains: [],
  });
  return (
    <Button onClick={() => { connect() }}>Connect</Button>
  );
};

```

**Reply by yutingzhao1991** [reply]
应该直接暴露 `ConnectModal` 来用：

```tsx
import { useConnector, ConnectModal } from '@ant-design/web3';

export default () => {
  const [wallet, setWallet] = React.useState();
  const { connect } = useConnector({
    chains: [],
  });
  return (
   <ConnectModal wallet={wallet} onWalletSelect={setWallet}>
      <Button onClick={() => { connect(wallet) }}>Connect</Button>
    </ConnectModal>
  );
};
```

**Reply by kiner-tang** [reply]
感觉这种方式或许会更好

**Comment by yutingzhao1991** [comment]
梳理了一下现在的想法，感觉可以分为三个组件 `Connector`, `ConnectButton`, `ConnectModal`。

其中 `ConnectButton` 和 `ConnectModal` 是基础的纯 UI 的组件，`Connector` 是不包含 UI 的，用于整合 `ConnectButton`, 和`ConnectModal` 的。

首先 `ConnectButton` 可以单独就作为一个 UI 组件来用：

```tsx
export default () => {
  return (
    <ConnectButton
      address={'0x1111....111'}
      domain={'xxx.eth'}
      connected={false}
      chains={[]}
      onClick={() => {}}
      banlance={[{
        amount: 21,
        type: 'ETH',
      }]}
      avatar={}
      menuItems={}
      walletIcon={}
      tooltip
    />
  );
};
```

`ConnectModal` 也是一样，基本和现在一致，用户完全可以只以 UI 的方式来使用，和用 antd 的 Button 和 Modal 差不多：

```tsx
import { useConnector, ConnectModal, ConnectButton } from '@ant-design/web3';

export default () => {
  const [wallet, setWallet] = React.useState();
  const { connect } = useConnector({
    chains: [],
  });
  return (
   <ConnectModal wallet={wallet} onWalletSelect={setWallet}>
      <ConnectButton onClick={() => { connect(wallet) }}>Connect</ConnectButton>
    </ConnectModal>
  );
};
```


但是我们再提供一个 `Connector` 组件，它的作用是把 `ConnectModal` 和 `ConnectButton` 整合，基于 `Web3ConfigProvider` 的 Context 来构造 `ConnectButton` 的 props，用法如下：

```tsx
export default () => {
  return (
    <Connector showBalance={['ETH', 'USDT']} showDomain modalProps={}>
      <ConnectButton avatar={}>
    </Connector>
  );
}
```


关键点是：
- ConnectButton 不调用链相关接口，只接收数据，负责展示。ConnectModal 也一样。
- Connector 负责按照用户传入的 props 调用链的接口，转换为 ConnectButton 和 ConnectModal 的 props 传递给它们。
- 用户也可以通过 `modalProps` 或者直接在 `ConnectButton` 来定义 props 来覆盖 Connector 提供的 props。


@kiner-tang @jeasonstudio 看看这个设计怎么样，感觉这样搞就比较清晰了。

**Reply by yutingzhao1991** [reply]
https://github.com/ant-design/ant-design-web3/issues/86 准备开搞了。
